tools/src/main/java/com/nedap/archie/archetypevalidator/ArchetypeValidator.java

```diff
@@ -238,4 +239,23 @@ private Archetype cloneAndPreprocess(MetaModels models, Archetype archetype) {
         return messages;
     }
 
+    private void checkForInfiniteLoopInSpecialisation(FullArchetypeRepository repository, Archetype archetype) {
+        List<String> archetypesInSpecialisationTree = new ArrayList<>();
+        archetypesInSpecialisationTree.add(archetype.getArchetypeId().getFullId());
+        Archetype childArchetype;
+        Archetype specialisationCheck = archetype;
+        while (specialisationCheck.getParentArchetypeId() != null) {
+            childArchetype = specialisationCheck;
+            specialisationCheck = repository.getArchetype(specialisationCheck.getParentArchetypeId());
+            if (specialisationCheck == null) {
+                // Parent cannot be found, will fail later in the validation
+                break;
+            }
+            if (archetypesInSpecialisationTree.contains(specialisationCheck.getArchetypeId().getFullId())) {
+                throw new UnsupportedOperationException("Infinite loop caused by specialising: " + specialisationCheck.getArchetypeId().getFullId() + " in " + childArchetype.getArchetypeId().getFullId());
```

Comment:
What Exception could best be thrown here?

Reply:
I think this should just add  message to the ValidationResult and fail validation for this archetype for this reason, rather than throwing an exception